### PR TITLE
feat(oidc): add pkce support

### PR DESCRIPTION
**Pull request by FineWolf**
Implementation for #2921 .

Two new options are added in the configuration schema:

- `enforce_pkce`
   String. Can either be `no`, `public_only`, or `yes`. Default is `public_only` (not a lot of clients supports PKCE when the code exchange is done in the backend).
- `enable_pkce_plain_challenge`
   Boolean. Defaults to `false`.

Documentation has been updated to reflect the new options.
`provider.go` has been modified to enable PKCE based on the new options.
Tests have been adjusted for the new options.

### CODE REVIEW

**Comment by james-d-elliott** [comment]
I believe we'll need to also add the metadata option code_challenge_methods_supported, which should include S256, and plain if the plain option is enabled. Unless I'm reading the spec wrong not including this would signal to clients that PKCE is not supported.


Excerpt from https://datatracker.ietf.org/doc/html/rfc8414#section-2:
```
code_challenge_methods_supported
      OPTIONAL.  JSON array containing a list of Proof Key for Code
      Exchange (PKCE) [[RFC7636](https://datatracker.ietf.org/doc/html/rfc7636)] code challenge methods supported by this
      authorization server.  Code challenge method values are used in
      the "code_challenge_method" parameter defined in [Section 4.3 of
      [RFC7636]](https://datatracker.ietf.org/doc/html/rfc7636#section-4.3).  The valid code challenge method values are those
      registered in the IANA "PKCE Code Challenge Methods" registry
      [[IANA.OAuth.Parameters](https://datatracker.ietf.org/doc/html/rfc8414#ref-IANA.OAuth.Parameters)].  If omitted, the authorization server
      does not support PKCE.
```

**Comment by FineWolf** [comment]
@james-d-elliott You are absolutely right, small oversight on my part.

PR has been adjusted to include the appropriate metadata.

**Review by clems4ever — COMMENTED** [review]
Thanks for this PR @FineWolf , it's all good for me in terms of code. And thanks for the test. Great job!

I only have one last question. I'm wondering if the values for this `enforce_pkce` option should not be called `all`, `public_only` and `none` instead of `yes`, `no` and `public_only`. Because when I read `yes` at the beginning I thought the type of the attribute was supposed to be a bool and so `true` instead of `yes` would make more sense but it's not (it's a string, an enum semantically) and therefore having a mix between `yes`, `no` and a third value did not make sense to me. I think all values of this option should reflect what this policy affects instead of sometimes stating what's it affects (public_only) and sometimes whether it affects anything (yes or no). Do you see what I mean?

**Comment by FineWolf** [comment]
> I only have one last question. I'm wondering if the values for this `enforce_pkce` option should not be called `all`, `public_only` and `none` instead of `yes`, `no` and `public_only`.

Indeed. Actually, `never`, `public_clients_only`, `always` is a bit more descriptive. I've updated the PR to use those values.

File: docs/configuration/identity-providers/oidc.md [hunk 1]
```diff
@@ -159,6 +159,7 @@ identity_providers:
     id_token_lifespan: 1h
     refresh_token_lifespan: 90m
     enable_client_debug_messages: false
+    enforce_pkce: public_clients_only
```
Comment (james-d-elliott) [inline docs/configuration/identity-providers/oidc.md:162]:
```suggestion
    enforce_pkce: always
```

File: config.template.yml [hunk 2]
```diff
@@ -684,6 +684,10 @@ notifier:
     ## security reasons.
     # minimum_parameter_entropy: 8
 
+    ## SECURITY NOTICE: It's not recommended changing this option, and highly discouraged to have it set to 'never'
+    ## for security reasons.
+    # enforce_pkce: public_clients_only
```
Comment (james-d-elliott) [inline config.template.yml:689]:
```suggestion
    # enforce_pkce: always
```

File: docs/configuration/identity-providers/oidc.md [hunk 3]
```diff
@@ -309,6 +310,39 @@ This controls the minimum length of the `nonce` and `state` parameters.
 make certain scenarios less secure. It is highly encouraged that if your OpenID Connect RP does not send these parameters
 or sends parameters with a lower length than the default that they implement a change rather than changing this value.
 
+### enforce_pkce
+<div markdown="1">
+type: string
+{: .label .label-config .label-purple } 
+default: public_only
```
Comment (james-d-elliott) [inline docs/configuration/identity-providers/oidc.md:317]:
```suggestion
default: always
```

File: docs/configuration/identity-providers/oidc.md [hunk 4]
```diff
@@ -309,6 +310,39 @@ This controls the minimum length of the `nonce` and `state` parameters.
...
+{: .label .label-config .label-blue }
+required: no
+{: .label .label-config .label-green }
+</div>
+
+[Proof Key for Code Exchange](https://datatracker.ietf.org/doc/html/rfc7636) enforcement policy: if specified, must be either `never`, `public_clients_only` or `always`.
+
+If set to `public_clients_only` (default), PKCE will be required for public clients using the Authorization Code flow.
+
+When set to `always`, PKCE will be required for all clients using the Authorization Code flow.
```
Comment (james-d-elliott) [inline docs/configuration/identity-providers/oidc.md:327]:
```suggestion
When set to the default of `always`, PKCE will be required for all clients using the Authorization Code flow.

If set to `public_clients_only`, PKCE will be required for public clients using the Authorization Code flow.
```

File: internal/configuration/config.template.yml [hunk 5]
```diff
@@ -684,6 +684,10 @@ notifier:
     ## security reasons.
     # minimum_parameter_entropy: 8
 
+    ## SECURITY NOTICE: It's not recommended changing this option, and highly discouraged to have it set to 'never'
+    ## for security reasons.
+    # enforce_pkce: public_clients_only
```
Comment (james-d-elliott) [inline internal/configuration/config.template.yml:689]:
```suggestion
    # enforce_pkce: always
```

File: internal/configuration/schema/identity_providers.go [hunk 6]
```diff
@@ -49,6 +52,7 @@ var DefaultOpenIDConnectConfiguration = OpenIDConnectConfiguration{
 	AuthorizeCodeLifespan: time.Minute,
 	IDTokenLifespan:       time.Hour,
 	RefreshTokenLifespan:  time.Minute * 90,
+	EnforcePKCE:           "public_clients_only",
```
Comment (james-d-elliott) [inline internal/configuration/schema/identity_providers.go:55]:
```suggestion
	EnforcePKCE:           "always",
```

File: internal/configuration/validator/const.go [hunk 7]
```diff
@@ -121,6 +121,9 @@ const (
 		"more clients configured"
 	errFmtOIDCNoPrivateKey = "identity_providers: oidc: option 'issuer_private_key' is required"
 
+	errFmtOIDCEnforcePKCEInvalidValue = "identity_providers: oidc: option 'enforce_pkce' must be 'never', " +
+		"'public_clients_only' or 'always', but it is configured as '%s'"
```
Comment (james-d-elliott) [inline internal/configuration/validator/const.go:125]: Nice job sticking to the implied style!

**Review by james-d-elliott — CHANGES_REQUESTED** [review]
LGTM! Few changes based on discussions with the team. We feel since it's configurable and OIDC is specifically a beta feature, we can make a potentially breaking change and have them change it as a preference to security over convenience.

In addition I was wondering what you thought about changing enable_pkce_plain_challenge to something like `pkce_code_challenge_methods` as a slice with the relevant options? I"m not hard set on it and it would require some additional validation logic, but felt like maybe it's a decent alternative in the event additional challenge modes become available over time.

**Comment by FineWolf** [comment]
I can definitely change `pkce_code_challenge_methods` to a slice with `S256` and `plain` as valid options. I won't be able to make those changes before Thursday however.

---

As for changing `enforce_pkce`'s default value to `always`, I would personally recommend against it.

Support for PKCE isn't widespread yet, and it was really meant to solve the issue where frontend/SPA/mobile applications couldn't use the `client_secret` as a reliable way of authenticating the application making the token request. Without PKCE, frontend/SPA/mobile applications were forced to use the implicit flow or the authorization code flow which were vulnerable to interception attacks.

In a situation where a client app is not a public app and has a `client_secret` to rely on to authenticate the source of the token request, PKCE at most an added countermeasure.

Nothing however is stopping private applications from requesting `PKCE` when `enforce_pkce` is set to `public_clients_only`; if a `code_challenge` is sent along with the authorization request, it will be taken into account and will enforce the token request to have the appropriate `code_verifier` along with it. However, it will allow authorization requests without them. The application however, because it is a private app, still needs to provide its `client_secret` to authenticate the source of the token request.

`public_clients_only` is a secure and sane default in my opinion. It still allows private clients to enroll in PKCE if they so desire, without outright rejecting the authorization request if no `code_challenge` is provided. Public apps using the auth code flow however are susceptible to authorization code interception attacks however, and have the most to gain by forcing applications to provide a PKCE challenge.

File: internal/configuration/config.template.yml (hunk 5 above)
Comment (FineWolf) [reply internal/configuration/config.template.yml:689]: See https://github.com/authelia/authelia/pull/2924#issuecomment-1056113203

File: docs/configuration/identity-providers/oidc.md (hunk 4 above)
Comment (FineWolf) [reply docs/configuration/identity-providers/oidc.md:327]: See https://github.com/authelia/authelia/pull/2924#issuecomment-1056113203

File: docs/configuration/identity-providers/oidc.md (hunk 3 above)
Comment (FineWolf) [reply docs/configuration/identity-providers/oidc.md:317]: See https://github.com/authelia/authelia/pull/2924#issuecomment-1056113203

Comment (FineWolf) [reply docs/configuration/identity-providers/oidc.md:317]: Also, comment for myself, that should have been `default: public_clients_only`.

File: config.template.yml (hunk 2 above)
Comment (FineWolf) [reply config.template.yml:689]: See https://github.com/authelia/authelia/pull/2924#issuecomment-1056113203

File: docs/configuration/identity-providers/oidc.md (hunk 1 above)
Comment (FineWolf) [reply docs/configuration/identity-providers/oidc.md:162]: See https://github.com/authelia/authelia/pull/2924#issuecomment-1056113203

File: docs/configuration/identity-providers/oidc.md (hunk 3 above)
Comment (james-d-elliott) [reply docs/configuration/identity-providers/oidc.md:192]: Oh Yep! I'll make this change quickly if it hasn't been done. Good spot!

**Comment by nightah** [comment]
@allcontributors please add @FineWolf for code, docs and tests.

**Pull request merged**